app/models/pusher.rb

```diff
@@ -93,7 +93,6 @@ def find
     sha256 = Digest::SHA2.base64digest(body.string)
 
     @version = @rubygem.versions.new number: spec.version.to_s,
-                                     canonical_number: spec.version.canonical_segments.join("."),
```

Comment:
I think the purpose of this explicit assignment is to ensure it is present once initialize (even currently no needed for pusher purpose and I wasn't able to find out any other place initializing new `Version`).

Anyway I was looking how this is used and I wasn't able to find the reason for keeping `canonical_number` and `number`. `canonical_number` seems used only for uniqueness validation, but on the model side only.  Real gem version uniqueness actually is covered by `index_versions_on_rubygem_id_and_number_and_platform` index. But for that one it seems "0.5.0" and "0.5" are different versions. So both approaches (model one by `canonical` and DB index) are enforcing uniqueness together (even the model one is always right, DB one could fail if somehow model ones passes).

I'll keep this in my notes to refactor a little and rely on `index` (with friendly model validation as well), both based on canonical and unique versions. For now it seems as not blocker. :thinking:

Reply:
The reason the canonical_number can't be added to the index is that, before the validation was added, there were multiple versions with the same canonical number in the DB.

Reply:
Thanks for the info. There are some techniques to exclude invalid data, I'll check.

Reply:
Just adding the PR that originally added this for context: https://github.com/rubygems/rubygems.org/pull/2559, including the discussion of how to normalize this data.

Reply:
Ahh, so we already been there. What a memory! It seems the following step was missed. :eyes: